pkg/authz/responsefilterer.go

```diff
@@ -0,0 +1,404 @@
+		negotiator := runtime.NewClientNegotiator(codecs, schema.GroupVersion{
+			Group:   gvk.Group,
+			Version: gvk.Version,
+		})
+		decoder, err := negotiator.Decoder(mediaType, params)
+		if err != nil {
+			return fmt.Errorf("failed to get decoder for %s: %w", mediaType, err)
+		}
+		encoder, err := negotiator.Encoder(contentType, params)
+		if err != nil {
+			return fmt.Errorf("failed to get encoder for %s: %w", contentType, err)
+		}
+
+		// If the object is proto encoded but not in the scheme, we can't decode it.
+		// Built-in types can be proto-encoded, custom types won't be, but perhaps
+		// this will change in the future - seeing this error on a new version of
+		// kube will require investigation to see what has changed with encoding.
+		if !recognized && strings.Contains(mediaType, "proto") {
+			return fmt.Errorf("unsupported media type %s for gvk %s", mediaType, gvk.String())
+		}
+
+		switch len(info.Parts) {
+		// if there's 1 part in the url (i.e. "pods"), we assume it's a list
+		case 1:
+			if !strings.HasSuffix(gvk.Kind, "List") {
+				gvk.Kind = gvk.Kind + "List"
+			}
+
+			var typedList runtime.Object
+
+			if recognized {
+				var err error
+				typedList, err = scheme.Scheme.New(gvk)
+				if err != nil {
+					return fmt.Errorf("failed to create new object of type %s: %w", gvk, err)
+				}
+			} else {
+				// custom types
+				typedList = &unstructured.UnstructuredList{}
+			}
+			_, _, err = decoder.Decode(body, &gvk, typedList)
+			if err != nil {
+				return fmt.Errorf("failed to decode response body: %w", err)
+			}
+			filterErr = rf.filterList(typedList, result)
+			if filterErr != nil {
+				break
+			}
+			if recognized {
+				filterErr = encoder.Encode(typedList, &filteredBody)
+			} else {
+				// custom types
+				filterErr = unstructured.UnstructuredJSONScheme.Encode(typedList, &filteredBody)
+			}
+			if filterErr != nil {
+				break
+			}
+
+		// if there's 2 or parts in the url (i.e. "pods/foo", "pods/foo/status"), we assume it's a single object
+		default:
+			var typedObj runtime.Object
+
+			if recognized {
+				var err error
+				typedObj, err = scheme.Scheme.New(gvk)
+				if err != nil {
+					return fmt.Errorf("failed to create new object of type %s: %w", gvk, err)
+				}
+			} else {
+				// custom types
+				typedObj = &unstructured.Unstructured{}
+			}
+			_, _, err = decoder.Decode(body, &gvk, typedObj)
+			if err != nil {
+				return fmt.Errorf("failed to decode response body: %w", err)
+			}
+
+			filterErr = rf.filterObject(typedObj, result)
+			if filterErr != nil {
+				break
+			}
+			filteredBody = *bytes.NewBuffer(body)
+		}
+
+		return writeResp(filteredBody, filterErr, resp)
+	}
+}
+
+func writeResp(filteredBody bytes.Buffer, filterErr error, resp *http.Response) error {
+	// if there was an error, replace the body with an error message
+	if filterErr != nil {
+		filteredBody.Reset()
+		var merr error
+		errBody, merr := kjson.Marshal(k8serrors.NewUnauthorized(filterErr.Error()))
```

Comment:
I understand this predates the PR. I think this makes transient errors difficult to troubleshoot. I'd rather return a 500 error, as I fear this may get confused with normal operation of the proxy.

Reply:
Yeah, I didn't want to change this behavior